### PR TITLE
BUGFIX: SD-530 Can't export insight with filter by string

### PR DESCRIPTION
**Pull request by ptquang86**
---

# PR checklist

- [x] Change was tested using dev-release in Analytical Designer and Dashboards (if applicable).
- [x] Change is described in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Migration guide (for major update) is written to [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [ ] Squash commits

# Related PRs
- gooddata-react-components: https://github.com/gooddata/gooddata-react-components/pull/1080
- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2352
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2125

# Related Jira tasks
- SD-530: https://jira.intgdc.com/browse/SD-530

### CODE REVIEW

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/312/ : SUCCESS in 1m 08s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/374/ : SUCCESS in 3m 41s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/313/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/375/ : SUCCESS in 1m 04s

File: test/report.test.ts [hunk 1]
```diff
@@ -27,6 +28,125 @@ describe("report", () => {
...
+                    title: "title",
+                    format: "xlsx",
+                    mergeHeaders: false,
+                    showFilters,
+                };
+
+                return mockedReportModule()
+                    .exportResult(projectId, executionResult, exportConfig, { pollStep: 1 })
+                    .then(() => {
+                        const [, settings] = fetchMock.lastCall(
```
Comment (naibinh) [inline test/report.test.ts]: I'm not sure this style is invalid `const [, settings]`, but please consider this `const [_, settings]`

Comment (ptquang86) [reply test/report.test.ts]: Hi @naibinh, Validator shows this error:

```
$ tsc --noEmit && yarn tslint && yarn prettier-check
test/report.test.ts:133:32 - error TS6133: '_' is declared but its value is never read.
133                         const [_, settings] = fetchMock.lastCall(
Found 1 error.
```

**Comment by naibinh** [comment]
LGTM

File: src/report.ts [hunk 2]
```diff
@@ -93,10 +100,19 @@ export class ReportModule {
         return (error as ApiResponseError).response !== undefined;
     }
 
-    private convertFilter(filter: AFM.CompatibilityFilter): AFM.CompatibilityFilter | null {
-        if ((filter as AFM.IExpressionFilter).value === undefined) {
-            return convertVisualizationObjectFilter(filter as VisualizationObjectFilter);
+    private sanitizeFilters(filter: AFM.CompatibilityFilter): ExecuteAFM.CompatibilityFilter | null {
```
Comment (ivan-nejezchleb) [inline src/report.ts]: this should be named sanitizeFilter (without s), because it transforms only one filter

File: src/report.ts [hunk 3]
```diff
@@ -47,16 +51,19 @@ export class ReportModule {
         exportConfig: IExportConfig = {},
         pollingOptions: IPollingOptions = {},
     ): Promise<IExportResponse> {
-        const showFilters: AFM.CompatibilityFilter[] = exportConfig.showFilters
-            ? compact(exportConfig.showFilters.map(this.convertFilter))
-            : [];
+        const { showFilters } = exportConfig;
+
+        let sanitizedFilters: ExecuteAFM.CompatibilityFilter[] | undefined;
```
Comment (ivan-nejezchleb) [inline src/report.ts]: move this to the separate function sanitizeFilters(showFilters: ExecuteAFM.CompatibilityFilter[]) {
        if (isArray(showFilters) && showFilters.length > 0) {
            return compact(showFilters.map(this.sanitizeFilter));
        }
        return undefined;
}

to split the code and get rid of 'let' :)

File: src/report.ts [hunk 4]
```diff
@@ -93,10 +100,19 @@ export class ReportModule {
...
 
-    private convertFilter(filter: AFM.CompatibilityFilter): AFM.CompatibilityFilter | null {
-        if ((filter as AFM.IExpressionFilter).value === undefined) {
-            return convertVisualizationObjectFilter(filter as VisualizationObjectFilter);
+    private sanitizeFilters(filter: AFM.CompatibilityFilter): ExecuteAFM.CompatibilityFilter | null {
+        if (AFM.isAttributeFilter(filter)) {
+            return convertAttributeFilter(filter);
+        } else if (AFM.isAbsoluteDateFilter(filter)) {
+            return convertAbsoluteDateFilter(
+                filter as VisualizationObject.IVisualizationObjectAbsoluteDateFilter,
```
Comment (ivan-nejezchleb) [inline src/report.ts]: we don't like these explicit changes of type. If these two structures are completely identical, only have two different names (types) than converter implementation should accept both of them as input param ->
`function convertRelativeDateFilter(
    filter: IVisualizationObjectRelativeDateFilter | AFM.CompatibilityFilter,
): AFM.FilterItem | null {`

File: src/report.ts (hunk 2 above)
Comment (ptquang86) [reply src/report.ts]: updated. thanks :)

File: src/report.ts (hunk 3 above)
Comment (ptquang86) [reply src/report.ts]: updated

File: src/report.ts (hunk 4 above)
Comment (ptquang86) [reply src/report.ts]: updated. thank you :)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/314/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/376/ : SUCCESS in 3m 31s

File: src/report.ts [hunk 5]
```diff
@@ -93,10 +96,24 @@ export class ReportModule {
...
+    }
+
+    private sanitizeFilter(filter: AFM.CompatibilityFilter): ExecuteAFM.CompatibilityFilter | null {
+        if (AFM.isAttributeFilter(filter)) {
+            return convertAttributeFilter(filter);
+        } else if (AFM.isAbsoluteDateFilter(filter)) {
+            return convertAbsoluteDateFilter(filter);
+        } else if (AFM.isRelativeDateFilter(filter)) {
+            return convertRelativeDateFilter(filter);
+        } else {
```
Comment (ivan-nejezchleb) [inline src/report.ts]: this else is not needed

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: src/report.ts (hunk 5 above)
Comment (ptquang86) [reply src/report.ts]: updated and squashed :)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/316/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/378/ : SUCCESS in 1m 46s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-merge-compliance-check/86/ : SUCCESS in 31s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/379/ : SUCCESS in 1m 57s

**Pull request merged**